✨ Progress bar improvements for large stories

Enriqe opened the pull request:
Closes #20175

Demo https://stamp-animation.firebaseapp.com/examples/amp-story/progress-bar-overflow.html

newmuis commented:
Without having looked at the code, the replay action in the bookend doesn't jump the progress bar back to the beginning.

It also looks like the last page of the story may have extra padding at the end of the progress bar?

gmajoulet commented:
Thanks for uploading the demo!
I've been playing with it for a bit on different devices, and unfortunately I don't think the animation meets the bar in terms of performance, or visually.

The fact that we animate on the `width` attribute makes everything very slow on regular devices like the Nexus 5, and still not smooth on an iPhone 7.

Visually there are a few things we could and should fix. For example, the segments are expanded by default, shrunk into an ellipsis as they enter the viewport, but seeing them entering while shrinking adds unnecessary noise on the animation, and makes little visual sense. Also, the segments should be displayed with no animation on page load, and sometimes the positioning is off ([only two ellipsis here](https://user-images.githubusercontent.com/1492044/63230553-2341ee00-c1dc-11e9-89a0-c77b0d42c53d.png)). We need to handle resize events as well.

I've been trying to think of a solution for the performance issue, and I believe it'd allow us both that and all the visual issues with much less complexity.

Here's what I had in mind:

- Each segment is an ellipsis, a 3px * 3px square
- To avoid animating the width, they're expanded using `scaleX(x)`
- Each segment is positioned using a `translate3d(x, y, z)` property
- `RTL` support gets extremely easy since all you have to do is invert the `translate3d()`

I wrote a prototype here to make sure it could work, and it ended up being less than 40 lines of logic. I believe we could turn this 500 lines PR into a ~100-200 lines one, while getting much better performance and easier code.

https://codepen.io/gmajoulet/pen/ExYyOrV

Let me know what you think, I know we've been talking about this idea before, but seeing how much easier, lightweight and performant this could become, I think it might be worth spending the extra time to refactor this PR.

Enriqe commented:
Thanks @gmajoulet

As you mentioned we initially agreed on animating the width of the segments and treating the progress bar as a "slide". After some testing in real world devices this showed to have poor performance. In hindsight I should have probably tested the prototype on a device before implementation.

I went ahead and changed the implementation so that every segment is an ellipse and only grow those ones that are in range to be shown to the user. The segments are individually grown with a `scale` and moved to their corresponding position with a `translate`.

Edge cases like reloading in the middle of the story, RTL, and pressing on the replay button are being handled as well. Haven't tested with live stories yet, though I think it should work (I'll get back on that).

Some open questions:
* What do we want to do for desktop 3 panels? 
  * Option A: Leave as is, showing both ellipses and segments to the user, but center it (right now it's a bit off-center in the demo).
  * Option B: Expand all the segments, since by nature desktop users have more real estate and probably don't have the problem that we are trying to solve in the first place.
  * Option C: Adapt the progress bar so that it looks like the one in mobile. It would be in the same position but it would follow the same logic.
* Live stories behavior [TODO]

Enriqe commented:
Btw the demo was updated with the new impl.

https://stamp-animation.firebaseapp.com/examples/amp-story/progress-bar-overflow.html

Enriqe commented:
FYI: When testing in a real world device I'm seeing some blinking when the animation happens. I'm looking into this

Enriqe commented:
Bug has been fixed and resize events are now being handled. Other than the open question in https://github.com/ampproject/amphtml/pull/23884#issuecomment-523221871 about the desktop panels this is ready for another review 👌🏽

gmajoulet commented on extensions/amp-story/1.0/amp-story-system-layer.css:
```diff
@@ -82,15 +82,17 @@
   float: left !important;
 }
 
+ol.i-amphtml-animate-progress li {
```
Please avoid using html tags when you can use class names

gmajoulet commented on extensions/amp-story/1.0/amp-story-system-layer.css:
```diff
@@ -82,15 +82,17 @@
 }
 
+ol.i-amphtml-animate-progress li {
+  transition: transform 1s cubic-bezier(0.4, 0.0, 0.2, 1) !important;
```
We'd need to check the animation speed with Hong, but I feel like a beet faster, like `.8s`, could be an ok place to start

gmajoulet commented on extensions/amp-story/1.0/amp-story-system-layer.css:
```diff
@@ -102,12 +104,17 @@
 
 .i-amphtml-story-page-progress-bar {
   background: rgba(255, 255, 255, 0.4) !important;
-  border-radius: 1px !important;
```
What can we do to keep this? I know it looks weird because of the `scale` transform, but is there anything we can do to keep a border radius?

Enriqe replied:
Sure, I just thought we wanted squared segments. But also, didn't we have performance issues due to using `border-radius` on resizing elements? I guess it doesn't apply here since we're not animating the width, thus not causing a repaint? (Maybe that a special case for videos in `amp-story-page` 3-panels desktop?) Just making sure. /cc @newmuis

Enriqe replied:
Also, the segments look weird with the radius on full-bleed with the `border-radius`:
![image](https://user-images.githubusercontent.com/5449100/63527926-fdfcfa80-c4cf-11e9-9d3f-f5365a10fdeb.png)

Enriqe replied:
Without `border-radius`:
![image](https://user-images.githubusercontent.com/5449100/63528382-d0648100-c4d0-11e9-8269-61172cb1da63.png)

gmajoulet replied:
Here's a screenshot of the original mock:
![image](https://user-images.githubusercontent.com/1492044/63528691-63052000-c4d1-11e9-89a5-42d0cad6e49c.png)

The `border-radius` issue we had on videos is unrelated here.
I know it looks weird because of the scaleX, but is there any way we can make it look good?

newmuis replied:
Just confirming: the border radius issues we had were all just because the video itself was being clipped.  This should be fine.

gmajoulet replied:
Resolution for this as seen with Hong: we keep the border radius when segmentsCount < MAX_SEGMENTS, and remove it when we trigger the overflow animation.

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -35,14 +37,41 @@ const TRANSITION_LINEAR = `transform ${POLL_INTERVAL_MS}ms linear`;
+ * Size in pixels of a segment ellipse.
+ * @const {number}
+ */
+const ELLIPSE_WIDTH_PX = 3;
```
If we want them to be circles, should they be 2px?

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -35,14 +37,41 @@ const TRANSITION_LINEAR = `transform ${POLL_INTERVAL_MS}ms linear`;
+ * Size in pixels of the total side margins of a segment.
+ * @const {number}
+ */
+const SEGMENTS_MARGIN_PX = 4;
```
Can we try a smaller margin for the ellipses and see if it looks better? It's hard to tell from the mocks.

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -72,14 +101,40 @@ export class ProgressBar {
+    this.segments_ = [];
+
+    /**
+     * Index next to first ellipse (if any) for stories with segments >
```
Does it mean it's the first expanded segment?

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -119,10 +177,180 @@ export class ProgressBar {
+  getPrevEllipsisCount_() {
+    return this.progressBarIndex_ > 3
+      ? 3
+      : Math.min(this.increment_, this.progressBarIndex_);
```
I think `return Math.min(3, this.progressBarIndex_);` would solve this?

This way, you could remove the `this.increment_` ivar and keep it as a local variable in `checkIndexForOverflow_`.

Enriqe replied:
Done.

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -119,10 +177,180 @@ export class ProgressBar {
+  }
+
+  /**
+   * Gets the count of ellipsis to the right (in LTR).
```
Plz update

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -119,10 +177,180 @@ export class ProgressBar {
+  }
+
+  /**
+   * Gets the count of ellipsis to the left (in LTR).
```
Plz update

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -119,10 +177,180 @@ export class ProgressBar {
+  }
+
+  /**
+   * Renders the segments by setting their corresponding width and translate.
```
Nit: s/width/scaleX

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -119,10 +177,180 @@ export class ProgressBar {
+      -(this.progressBarIndex_ - this.getPrevEllipsisCount_()) *
+      (ELLIPSE_WIDTH_PX + SEGMENTS_MARGIN_PX);
+
+    for (let index = 0; index < this.segmentCount_; index++) {
```
Can we wrap this entire loop in a `mutate` to avoid the overhead of calling `mutate` for each segment?

The big advantage is that you can get a promise out of `mutate`, that will resolve when the mutation has been executed.
This will be a useful heuristic to know when to add or not the `animate` class you have. You could think of adding a `shouldAnimate` optional parameter, and then toggle the class before/after the animation as needed.
It'd be great so we could remove the animation on pageload, resize, replay.

WDYT?

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -119,10 +177,180 @@ export class ProgressBar {
+      // Don't animate the progress bar in pageload.
+      this.timer_.delay(() => {
+        this.root_.classList.add('i-amphtml-animate-progress');
+      }, 60);
```
60ms seems very arbitrary, we should remove this, cf the other comment on the `render_` method.

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -193,6 +422,12 @@ export class ProgressBar {
 
     this.updateProgressByIndex_(segmentIndex, progress);
 
+    // If story was reloaded.
```
Isn't this also called on initial render, even on the cover page?
Could we rename the `updateValuesForReload_` method so it better indicates what it does, like `getInitialSegmentIndex_`?

Enriqe replied:
Done.

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -207,6 +442,22 @@ export class ProgressBar {
+  /**
+   * Snap the progressBarIndex to its most appropiate place after reload.
+   * @param {number} segmentIndex
+   */
```
Nit: `@private`

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -119,10 +180,191 @@ export class ProgressBar {
+        this.root_.classList.toggle(
+          'i-amphtml-animate-progress',
+          shouldAnimate
+        );
```
Does this work? I'm a bit confused since you're toggling the class after setting the transform properties.

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -119,10 +180,191 @@ export class ProgressBar {
+   * @private
+   */
+  checkIndexForOverflow_() {
+    // Touching an ellipsis on the right (LTR).
```
Plz update this comment and the following one. We don't need to talk about right or left but simply previous/next, which work for both RTL and LTR

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -193,6 +436,11 @@ export class ProgressBar {
 
     this.updateProgressByIndex_(segmentIndex, progress);
 
+    if (!this.activeSegmentIndex_) {
```
Sorry I missed this in the first review: is there any reason why we want to position the segments for the first time in the first call to `updateProgress()` and not in the `build()`?

gmajoulet commented on extensions/amp-story/1.0/progress-bar.js:
```diff
@@ -35,14 +37,47 @@ const TRANSITION_LINEAR = `transform ${POLL_INTERVAL_MS}ms linear`;
+ * Size in pixels of a segment ellipse.
+ * @const {number}
+ */
+const ELLIPSE_WIDTH_PX = 2;
```
Yep sorry they should be `2px` on mobile and `3px` on both desktop versions, to reflect our CSS

Enriqe commented on extensions/amp-story/1.0/amp-story-system-layer.css:
```diff
@@ -315,19 +319,43 @@
   height: 3px !important;
-  width: calc(100vw / 3) !important;
+  width: 100% !important;
+  max-width: calc(3/5 * 75vh) !important;
```
I copied these styles from the `amp-story-page` dimensions in the desktop panels css file so that the widths match:

https://github.com/ampproject/amphtml/blob/94227379140566c174cec3ac7f27c2838ce45e9a/extensions/amp-story/1.0/amp-story-desktop-panels.css#L122-L132

newmuis replied:
Can't you just use `max-width: 100%`?

Enriqe replied:
Hmmm not sure I understand what you're implying. We need the `max-width` to match in both the `amp-story-page` and the `progress-bar` in the desktop panels. 

Right now, the `max-width` for the `amp-story-page` is `max-width: calc(3/5 * 75vh`, so I just changed the progress bar width in desktop panels to match that.

newmuis replied:
Whoops, sorry, forgot that the progress bar is outside the page 😅

newmuis replied:
In this case, I wonder: can you use `max-width: calc(100 * --story-page-vw)`?

newmuis replied:
Ugh, actually that will cause weird jank on resize, since the custom property gets updated through the (debounced) JS resize handler.

I guess you can just do what you wrote here.

gmajoulet replied:
Could we keep the current design (not copy any CSS), and measure the progress bar element instead of measuring the page?
I don't think there is any need to change the design to write this animation.

Enriqe commented:
@newmuis If we add padding then more than 3 ellipsis will be shown. I've tried a couple of things but didn't find an easy way that worked across screens (desktop, mobile, fullbleed). If you feel strongly I can fix in a follow-up.

Enriqe commented:
Updated the demo and addressed feedback. PTAL :)

gmajoulet commented on extensions/amp-story/1.0/amp-story.js:
```diff
@@ -1041,21 +1038,25 @@ export class AmpStory extends AMP.BaseElement {
+    const maybeHistoryPage = historyPage
+      ? this.element.querySelector(`#${escapeCssSelectorIdent(historyPage)}`)
+      : null;
+    if (historyPage && maybeHistoryPage) {
```
Not sure why you removed the function that factorized that code, we could simply update the `isActualPage` function and keep the current implementation
```
const isActualPage = pageId => !!this.element.querySelector(`#${escapeCssSelectorIdent(historyPage)}`);
```

Enriqe replied:
Yep my bad. Done.

gmajoulet commented on extensions/amp-story/1.0/amp-story-system-layer.css:
```diff
@@ -315,11 +315,22 @@
+
+.i-amphtml-progress-bar-overflow.i-amphtml-story-progress-bar {
+  width: 100% !important;
+  margin: 4px 0 0 !important;
```
Keeping the borders as asked by @newmuis should be as simple as:

```
width: calc(100% - 4px);
margin: 4px 2px 0;
overflow: hidden;
```

Enriqe replied:
🤦🏽‍♂️that's easy (i was trying to do it with padding). thanks!

gmajoulet replied:
Everything got confusing because we all change the box model and set box-sizing to border-box, but margin are outside the content, and padding inside the content :))

gmajoulet commented on extensions/amp-story/1.0/amp-story-system-layer.css:
```diff
@@ -315,11 +315,22 @@
   height: 3px !important;
   width: calc(100vw / 3) !important;
   margin: 0 24px !important;
+  overflow: hidden !important;
```
Plz remove if you add it from the comment above
